README
List-BinarySearch

Perform a binary search on a flat list or an array passed by reference.

v0.10 adds bsearch_custom_range function, and POD description of 
Unicode support.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

There are no non-core dependencies.
This module a minimum Perl version of 5.6.0.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc List::BinarySearch

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=List-BinarySearch

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/List-BinarySearch

    CPAN Ratings
        http://cpanratings.perl.org/d/List-BinarySearch

    Search CPAN
        http://search.cpan.org/dist/List-BinarySearch/


LICENSE AND COPYRIGHT

Copyright (C) 2012 David Oswald

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for List::BinarySearch

0.24 2014-12-02
  - Fix SYNOPSIS typos in L::BS and L::BS::PP (Closes GH issue #1).

0.23 2014-12-01
  - Remove some diagnostics code.

0.22 2014-12-01
  - Comment reason why we goto &Exporter::import rather than calling SUPER::.
  - Make it easier to manage distribution version numbering.
  - Shift bugs/feature requests from RT to GitHub issues.
  - Recommend List::BinarySearch::XS v0.06 (was v0.05).

0.21 2014-12-01
  - Improve logic for overriding 'import' on pre-Perl 5.20.

0.20 2014-02-20
  - This time really do fix version numbers. ;)

0.18 2014-02-05
  - Fix version numbers.
  
0.17 2013-11-05
  - Custom wrapper for import() that touches $a and $b in caller's
    namespace for Perl versions < 5.020, to squelch "used only
    once" warnings.  This issue will be fixed in Perl 5.20.

0.16 2013-10-28
  - Fixed POD typo (bug in example).

0.15 2013-09-27
  - Removed unnecessary max==min test in L::BS::PP::binsearch (save a cycle:
    just an efficiency improvement).

0.14 2013-09-26
  - NEW API IS NOW STABLE: DEPRECATED FUNCTIONS HAVE BEEN REMOVED.
  - List::BinarySearch::XS will now be used as an optional default dependency.
  - In interactive mode, CPAN installers will prompt for permission to use 
    this dependency.  Default action is yes.
  - Set LBS_NO_XS to prevent installing the XS module, or answer 'no' in
    installer's interactive mode.
  - Test coverage is now 100%.

0.13_001 2013-09-24
  - DEVELOPERS RELEASE -- Testing Makefile's optional dependency code.
  - Added meta spec version and git repo to META information.
  - Updated examples/bsearch.pl to use binsearch
  - Makefile.PL prompts to add List::BinarySearch::XS dependency.
    (In non-interactive environment, this defaults to yes.)
    See POD for additional configuration information.
  - META_MERGE => recommends => List::BinarySearch::XS, for those clients
    that understand the 'recommends' semantics from META spec 2.0.

0.13 2013-09-17
  - TRANSITIONAL PUBLIC RELEASE -- DEPRECATED FUNCTIONS WILL BE REMOVED SOON!
  - Minimum Perl version supported is now 5.8.  Earlier Perl versions weren't
    compatible with this module's use of prototypes.
  - Additional coverage tests added: 100% branch and 98% conditional coverage
    of non-deprecated functions. (Don't worry about a single untestable
    condition in non-critical code.)
  - Fixed version number inconsistency (while bumping to 0.13).

0.12 2013-09-05
  - TRANSITIONAL PUBLIC RELEASE
  - Many of the old functions have been deprecated and should be removed from
    user code in favor of the new binsearch, binsearch_pos, or binsearch_range
    functions, which are easier to use, faster, and will cure acne.
  - POD revisions.
  - Establish interoperability with List::BinarySearch::XS; Install this
    module, install List::BinarySearch::XS, and when you use List::BinarSearch
    it will automatically upgrade to the XS functions for high performance if
    they're available.
  - THIS IS YOUR ONLY WARNING: Version 0.20 will remove deprecated functions;
    see the POD for tips on simple migration to the new syntax.

0.11_003 2013-09-04
  - DEVELOPER'S RELEASE
  - Strengthened language on deprecation of obsolete functions.
  - Improvements to isa tests.
  - Fixed a typo in the POD demonstrating comparator complexity for complex
    datastructures (see, it bit me too).
  - Restore Perl 5.6 compatibility for the pure-Perl implementation.

0.11_002 2013-09-03
  - DEVELOPER'S RELEASE
  - POD revisions to recommend installing List::BinarySearch::XS
  - Updates to tests, including tests to verify we get the correct
    binsearch implementation depending on availability of XS module.
  - Split pure-Perl implementations of binsearch and binsearch_pos
    into List::BinarySearch::PP, included with distribution.
  - List::BinarySearch will now use List::BinarySearch::XS for 
    binsearch and binsearch_pos implementations, or the ::PP module
    (included) if the ::XS hasn't been installed.
    (May be overridden by setting $ENV{List_BinarySearch_PP} true.)

0.11_001 2013-08-20
  - DEVELOPER'S RELEASE
  - Deprecated all existing functions under their current names.
  - Three new functions: binsearch, binsearch_pos, binsearch_range. These
    replace bsearch_custom, bsearch_custom_pos, and bsearch_custom_range.
    The others will not be replaced, as they can be created using the more
    generalized functions.
  - Major documentation changes.
  - Announce upcoming List::BinarySearch::XS that when installed will be a
    silent and automatic performance upgrade.
  - $a and $b now permitted in place of $_[0] and $_[1].  The old @_ paramater
    passing has been deprecated.
  - The next major release, v0.12 will formally deprecate according to
    the list above.

0.11 2012-11-02
  - "Changes" now CPAN::Changes::Spec compliant.
  - t/09-changes.t added to verify spec compliance.

0.10 2012-10-25
  - API ADDITION: Added bsearch_custom_range function.
  - Added explanation and example to POD on how to handle lists 
    sorted according to the Unicode Collation Algorithm 
    (Unicode::Collate).
  - Added tests for bsearch_custom_range.
      
0.09 2012-05-30
  - BinarySearch.pm is now B::Lint friendly.
  - Added an examples/ directory.

0.08 2012-05-09
  - API IS STABLE.
  - Applied improvements to Makefile.PL from lessons learned in other
    more cranky modules.
  - Improved Kwalitee for Makefile.PL
  - Improved minimum Perl version handling (still 5.6.0).
  - Added t/13-kwalitee.t
  - Improved Perl::Critic compliance in test suite and Makefile.PL.
  - Minor bugfixes t/00-boilerplate.t

0.07 2012-05-05
  - API ADDITION: Added bsearch_custom_pos function.
  - API DELETION: Removed bsearch_general, and bsearch_general_range.
    They were redundant, and prone to user-error.
  - POD improvement: Added a quick reference to aid in selecting the
    most appropriate search function from the several choices.
  - Added bsearch_custom_pos.

lib/List/BinarySearch.pm
## no critic (RCS,prototypes)

package List::BinarySearch;

use 5.008000;
use strict;
use warnings;
use Carp;

use Scalar::Util qw( looks_like_number );


BEGIN {

  my @imports = qw( binsearch binsearch_pos );

  # Import XS by default, pure-Perl if XS is unavailable, or if
  # $ENV{List_BinarySearch_PP} is set.

  # This conditional has been tested manually.  Can't be automatically tested.
  # uncoverable condition right false
  if (
       $ENV{List_BinarySearch_PP}
    || ! eval 'use List::BinarySearch::XS @imports; 1;'  ## no critic (eval)
  ) {
    eval 'use List::BinarySearch::PP  @imports;';        ## no critic (eval)
  }

}

require Exporter;

our @ISA       = qw(Exporter);    ## no critic (ISA)

# Note: binsearch and binsearch_pos come from List::BinarySearch::PP
our @EXPORT_OK = qw(  binsearch         binsearch_pos       binsearch_range  );

our %EXPORT_TAGS = ( all => \@EXPORT_OK );

# The prototyping gives List::BinarySearch a similar feel to List::Util,
# and List::MoreUtils.

our $VERSION = '0.24';

# Needed for developer's releases: See perlmodstyle.
# $VERSION = eval $VERSION;    ## no critic (eval,version)


# Custom import() to touch $a and $b in Perl version < 5.20, to eliminate
# "used only once" warnings.
{
  if( $] < 5.020 ) {
    *import = sub {
      my $pkg = caller;
      no strict 'refs'; ## no critic(strict)
      ${"${pkg}::a"} = ${"${pkg}::a"};
      ${"${pkg}::b"} = ${"${pkg}::b"};
      # It would feel nicer to call shift->SUPER::import(@_), but 
      # Exporter::import appears to be too fragile for this type of wrapper.
      goto &Exporter::import;
    };
  }
}



# binsearch and binsearch_pos will be loaded from List::BinarySearch::PP or
# List::BinarySearch::XS.



sub binsearch_range (&$$\@) {
  my( $code, $low_target, $high_target, $aref ) = @_;
  my( $index_low, $index_high );

  # Forward along the caller's $a and $b.
  local( *a, *b ) = do{
    no strict 'refs';  ## no critic (strict)
    my $pkg = caller();
    ( *{$pkg.'::a'}, *{$pkg.'::b'} );
  };
  $index_low  = binsearch_pos( \&$code, $low_target,  @$aref );
  $index_high = binsearch_pos( \&$code, $high_target, @$aref );
  local( $a, $b ) = ( $aref->[$index_high], $high_target ); # Use our own.
  if(  $index_high == scalar @$aref    or    $code->( $a, $b ) > 0  )
  {
    $index_high--;
  }
  return ( $index_low, $index_high );
}



1;    # End of List::BinarySearch

__END__

=head1 NAME

List::BinarySearch - Binary Search within a sorted array.

=head1 SYNOPSIS

This module performs a binary search on an array.

Examples:


    use List::BinarySearch qw( :all );  # ... or ...
    use List::BinarySearch qw( binsearch  binsearch_pos  binsearch_range );


    # Some ordered arrays to search within.
    @num_array =   ( 100, 200, 300, 400, 500 );
    @str_array = qw( Bach Beethoven Brahms Mozart Schubert );


    # Find the lowest index of a matching element.

    $index = binsearch {$a <=> $b} 300, @num_array;
    $index = binsearch {$a cmp $b} 'Mozart', @str_array;      # Stringy cmp.
    $index = binsearch {$a <=> $b} 42, @num_array;            # not found: undef


    # Find the lowest index of a matching element, or best insert point.

    $index = binsearch_pos {$a cmp $b} 'Chopin', @str_array;  # Insert at [3].
    $index = binsearch_pos {$a <=> $b} 600, @num_array;       # Insert at [5].

    splice @num_array, $index, 0, 600
      if( $num_array[$index] != 600 );                        # Insertion at [5]

    $index = binsearch_pos { $a <=> $b } 200, @num_array;     # Matched at [1].


    # The following functions return an inclusive range.

    my( $low_ix, $high_ix )
        = binsearch_range { $a cmp $b } 'Beethoven', 'Mozart', @str_array;
        # Returns ( 1, 3 ), meaning ( 1 .. 3 ).

    my( $low_ix, $high_ix )
        = binsearch_range { $a <=> $b } 200, 400, @num_array;



=head1 DESCRIPTION

A binary search searches B<sorted> lists using a divide and conquer technique.
On each iteration the search domain is cut in half, until the result is found.
The computational complexity of a binary search is O(log n).

The binary search algorithm implemented in this module is known as a
I<Deferred Detection> variant on the traditional Binary Search.  Deferred
Detection provides B<stable searches>.  Stable binary search algorithms have
the following characteristics, contrasted with their unstable binary search
cousins:

=over 4

=item * In the case of non-unique keys, a stable binary search will always
return the lowest-indexed matching element.  An unstable binary search would
return the first one found, which may not be the chronological first.

=item * Best and worst case time complexity is always O(log n).  Unstable
searches may stop once the target is found, but in the worst case are still
O(log n).  In practical terms, this difference is usually not meaningful.

=item * Stable binary searches only require one relational comparison of a
given pair of data elements per iteration, where unstable binary searches
require two comparisons per iteration.

=item * The net result is that although an unstable binary search might have
better "best case" performance, the fact that a stable binary search gets away
with fewer comparisons per iteration gives it better performance in the worst
case, and approximately equal performance in the average case. By trading away
slightly better "best case" performance, the stable search gains the guarantee
that the element found will always be the lowest-indexed element in a range of
non-unique keys.

=back

B<< This module has a companion "XS" module: L<List::BinarySearch::XS> which
users are strongly encouraged to install as well. >>  If List::BinarySearch::XS
is also installed, C<binsearch> and C<binsearch_pos> will use XS code.  This
behavior may be overridden by setting C<$ENV{List_BinarySearch_PP}> to a
true value.  Most CPAN installers will either automatically install the XS
module, or prompt to automatically install it.  See CONFIGURATION for details.


=head1 RATIONALE

B<A binary search is pretty simple, right?  Why use a module for this?>

Quoting from
L<Wikipedia|http://en.wikipedia.org/wiki/Binary_search_algorithm>:  I<When Jon
Bentley assigned it as a problem in a course for professional
programmers, he found that an astounding ninety percent failed to code a
binary search correctly after several hours of working on it, and another
study shows that accurate code for it is only found in five out of twenty
textbooks. Furthermore, Bentley's own implementation of binary search,
published in his 1986 book Programming Pearls, contains an error that remained
undetected for over twenty years.>

So to answer the question, you might use a module so that you
don't have to write, test, and debug your own implementation.


B<< Perl has C<grep>, hashes, and other alternatives, right? >>

Yes, before using this module the user should weigh the other options such as
linear searches ( C<grep> or C<List::Util::first> ), or hash based searches. A
binary search requires an ordered list, so one must weigh the cost of sorting or
maintaining the list in sorted order.  Ordered lists have O(n) time complexity
for inserts.  Binary Searches are best when the data set is already ordered, or
will be searched enough times to justify the cost of an initial sort.

There are cases where a binary search may be an excellent choice. Finding the
first matching element in a list of 1,000,000 items with a linear search would
have a worst-case of 1,000,000 iterations, whereas the worst case for a binary
search of 1,000,000 elements is about 20 iterations.  In fact, if many lookups
will be performed on a seldom-changed list, the savings of O(log n) lookups may
outweigh the cost of sorting or performing occasional linear time inserts.


=head1 EXPORT

Nothing is exported by default.  C<binsearch>, C<binsearch_pos>, and
C<binsearch_range> may be exported by listing them on the export list.

Or import all functions by specifying C<:all>.

=head1 SUBROUTINES/METHODS

=head2 WHICH SEARCH ROUTINE TO USE

=over 4

=item * C<binsearch>: Returns lowest index where match is found, or undef.

=item * C<binsearch_pos>: Returns lowest index where match is found, or the
index of the best insert point for needle if the needle isn't found.

=item * C<binsearch_range>: Performs a search for both low and high needles.
Returns a pair of indices refering to a range of elements corresponding to
low and high needles.

=back

=head2 binsearch CODE NEEDLE ARRAY_HAYSTACK

    $first_found_ix = binsearch { $a cmp $b } $needle, @haystack;

Pass a code block, a search target, and an array to search.  Uses
the supplied code block C<$needle> to test the needle against elements
in C<@haystack>.

See the section entitled B<The Callback Block>, below, for an explanation
of how the comparator works
(hint: very similar to C<< sort { $a <=> $b } ... >> ).

Return value will be the lowest index of an element that matches target, or
undef if target isn't found.

=head2 binsearch_pos CODE NEEDLE ARRAY_HAYSTACK

    $first_found_ix = binsearch_pos { $a cmp $b } $needle, @haystack;

The only difference between this function and C<binsearch> is its return
value upon failure.  C<binsearch> returns undef upon failure.
C<binsearch_pos> returns the index of a valid insert point for
C<$needle>.

Pass a code block, a search target, and an array to search.  Uses
the code block to test C<$needle> against elements in C<@haystack>.

Return value is the index of the first element equaling C<$needle>.  If no
element is found, the best insert-point for C<$needle> is returned.


=head2 binsearch_range CODE LOW_NEEDLE HIGH_NEEDLE ARRAY_HAYSTACK

    my( $low, $high )
      = binsearch_range { $a <=> $b }, $low_needle, $high_needle, @haystack;

Given C<$low_needle> and C<$high_needle>, returns a set of indices that
represent the range of elements fitting within C<$low_needle> and
C<$high_needle>'s bounds.  This might be useful, for example, in finding all
transations that occurred between 02012013 and 02292013.

I<This algorithm was adapted from Mastering Algorithms with Perl, page 172 and
173.>

=head2 The callback block (The comparator)

Comparators in L<List::BinarySearch> are used to compare the target (needle)
with individual haystack elements, returning the result of the relational
comparison of the two values.  A good example would be the code block in a
C<sort> function.

Basic comparators might be defined like this:

    # Numeric comparisons:
    binsearch { $a <=> $b } $needle, @haystack;

    # Stringwise comparisons:
    binsearch { $a cmp $b } $needle, @haystack;

    # Unicode Collation Algorithm comparisons
    $Collator = Unicode::Collate->new;
    binsearch { $Collator->( $a, $b ) } $needle, @haystack;

C<$a> represents the target.  C<$b> represents the contents of the haystack
element being tested.  This leads to an asymmetry that might be prone to
"gotchas" when writing custom comparators for searching complex data structures.
As an example, consider the following data structure:

    my @structure = (
        [ 100, 'ape'  ],
        [ 200, 'cat'  ],
        [ 300, 'dog'  ],
        [ 400, 'frog' ]
    );

A numeric custom comparator for such a data structure would look like this:

    sub{ $a <=> $b->[0] }

In this regard, the callback is unlike C<sort>, because C<sort> is always
comparing to elements, whereas C<binsearch> is comparing a target with an
element.

Just as with C<sort>, the comparator must return -1, 0, or 1 to signify "less
than", "equal to", or "greater than".


=head1 DATA SET REQUIREMENTS

A well written general algorithm should place as few demands on its data as
practical.  The requirements that these Binary Search algorithms impose are:

=over 4

=item * B<The list must be in ascending sorted order>.

This is a big one.  The best sort routines run in O(n log n) time.  It makes no
sense to sort a list in O(n log n) time, and then perform a single O(log n)
binary search when List::Util C<first> could accomplish the same thing in O(n)
time without sorting.

=item * B<The list must be in ascending sorted order.>

A Binary Search consumes O(log n) time. We don't want to waste linear time
verifying the list is sordted, so B<there is no validity checking. You have
been warned.>

=item * B<These functions are prototyped> as (&$\@) or ($\@).

What this implementation detail means is that C<@haystack> is implicitly passed
by reference.  This is the price we pay for a familiar user interface, cleaner
calling syntax, and the automatic efficiency of pass-by-reference.

=item * B<Objects in the search lists must be capable of being evaluated for
relationaity.>

I threw that in for C++ folks who have spent some time with Effective STL.  For
everyone else don't worry; if you know how to C<sort> you know how to
C<binsearch>.

=back

=head1 UNICODE SUPPORT

Lists sorted according to the Unicode Collation Algorithm must be searched using
the same Unicode Collation Algorithm, Here's an example using
L<Unicode::Collate>'s C<< $Collator->cmp($a,$b) >>:

    my $found_index = binsearch { $Collator->cmp($a, $b) } $needle, @haystack;


=head1 CONFIGURATION AND ENVIRONMENT

List::BinarySearch is a thin layer that will attempt to load
List::BinarySearch::XS first, and if that module is unavailable, will fall back
to List::BinarySearch::PP which is provided with this distribution.

Most CPAN installers will automatically pull in and install the XS plugin 
for this module.  If in interactive mode, they will prompt first. To override 
default behavior, forcing CPAN installers to I<not> pull in the XS module, 
set the environment variable LBS_NO_XS true prior to installation.  There
will be no prompt, and the XS module won't be pulled in and installed.
However, the List::BinarySearch::XS plugin is strongly recommended, and should
only be skipped in environments where XS modules cannot be compiled.

If one wishes to install the XS module beforhand, or at any time later on, just
installing it in the usual fashion is sufficient for List::BinarySearch to
recognize and start using it.

By installing L<List::BinarySearch::XS>, the pure-Perl versions of C<binsearch>
and C<binsearch_pos> will be automatically replaced with XS versions for
markedly improved performance.  C<binsearch_range> also benefits from the XS
plug-in, since internally it makes calls to C<binsearch_pos>.

Users are strongly advised to install L<List::BinarySearch::XS>.  If, after
installing List::BinarySearch::XS, one wishes to disable the XS plugin, setting
C<$ENV{List_BinarySearch_PP}> to a true value will prevent the XS module from
being used by L<List::BinarySearch>.  This setting will have no effect on users
who use List::BinarySearch::XS directly.

For the sake of code portability, it is recommended to use List::BinarySearch 
as the front-end, as it will automatically and portably downgrade to the 
pure-Perl version if the XS module can't be loaded.


=head1 DEPENDENCIES

This module uses L<Exporter|Exporter>, and automatically makes use of
L<List::BinarySearch::XS> if it's installed on the user's system.

This module will attempt to install List::BinarySearch::XS unless the
environment variable C<LBS_NO_XS> is set prior to install, or if in interactive
mode, the user opts to skip this recommended step.

This module supports Perl versions 5.8 and newer.
The optional XS extension also supports Perl 5.8 and newer.


=head1 INCOMPATIBILITIES

This module is incompatible with Perl versions prior to 5.8.  In particular,
its use of prototypes isn't compatible with Perl 5.6 or older.  It would be
easy to eliminate the use of prototypes, but doing so would change calling
syntax.


=head1 DIAGNOSTICS


=head1 SEE ALSO

L<List::BinarySearch::XS>: An XS plugin for this module; install it, and this
module will use it automatically for a nice performance improvement.  May also
be used on its own.

=head1 AUTHOR

David Oswald, C<< <davido at cpan.org> >>

If the documentation fails to answer a question, or if you have a comment or 
suggestion, send me an email.


=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to
L<https://github.com/daoswald/List-BinarySearch/issues>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.



=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc List::BinarySearch

This module is maintained in a public repo at Github.  You may look for
information at:

=over 4

=item * Github: Development is hosted on Github at:

L<http://www.github.com/daoswald/List-BinarySearch>

=item * GitHub Issue tracker (report bugs here)

L<https://github.com/daoswald/List-BinarySearch/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/List-BinarySearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/List-BinarySearch>

=item * Search CPAN

L<http://search.cpan.org/dist/List-BinarySearch/>

=back



=head1 ACKNOWLEDGEMENTS

Thank-you to those who provided advice on user interface and XS
interoperability.

L<Mastering Algorithms with Perl|http://shop.oreilly.com/product/9781565923980.do>,
from L<O'Reilly|http://www.oreilly.com>: for the inspiration (and much of the
code) behind the positional and ranged searches.  Quoting MAwP: "I<...the
binary search was first documented in 1946 but the first algorithm that worked
for all sizes of array was not published until 1962.>" (A summary of a passage
from Knuth: Sorting and Searching, 6.2.1.)

I<Although the basic idea of binary search is comparatively straightforward,
the details can be surprisingly tricky...>  -- Donald Knuth


=head1 LICENSE AND COPYRIGHT

Copyright 2012 David Oswald.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


lib/List/BinarySearch/PP.pm
## no critic (RCS,prototypes)

package List::BinarySearch::PP;

use 5.006000;

use strict;
use warnings;
use Carp;

require Exporter;

our @ISA    = qw(Exporter);    ## no critic (ISA)
our @EXPORT = qw( binsearch binsearch_pos ); ## no critic (export)


our $VERSION = '0.24';
# $VERSION = eval $VERSION;  ## no critic (eval)



#---------------------------------------------
# Use a callback for comparisons.

sub binsearch (&$\@) {
    my ( $code, $target, $aref ) = @_;
    my $min = 0;
    my $max = $#{$aref};
    while ( $max > $min ) {
        my $mid = int( ( $min + $max ) / 2 );
        no strict 'refs'; ## no critic(strict)
        local ( ${caller() . '::a'}, ${caller() . '::b'} )
          = ( $target, $aref->[$mid] );
        if ( $code->( $target, $aref->[$mid] ) > 0 ) {
            $min = $mid + 1;
        }
        else {
            $max = $mid;
        }
    }
    {
      no strict 'refs'; ## no critic(strict)
      local ( ${caller() . '::a'}, ${caller() . '::b'} )
        = ( $target, $aref->[$min] );
      return $min if $code->( $target, $aref->[$min] ) == 0;
    }
    return;    # Undef in scalar context, empty list in list context.
}


#------------------------------------------------------
# Identical to binsearch, but upon match-failure returns best insert
# position for $target.


sub binsearch_pos (&$\@) {
    my ( $comp, $target, $aref ) = @_;
    my ( $low, $high ) = ( 0, scalar @{$aref} );
    while ( $low < $high ) {
        my $cur = int( ( $low + $high ) / 2 );
        no strict 'refs';  ## no critic(strict)
        local ( ${ caller() . '::a'}, ${ caller() . '::b'} )
          = ( $target, $aref->[$cur] );                            # Future use.
        if ( $comp->( $target, $aref->[$cur] ) > 0 ) {
            $low = $cur + 1;
        }
        else {
            $high = $cur;
        }
    }
    return $low;
}


1;

__END__

=head1 NAME

List::BinarySearch::PP - Pure-Perl Binary Search functions.


=head1 SYNOPSIS

This module is a plugin for List::BinarySearch providing a graceful fallback to
a pure-Perl binary search implementation in case the optional (but default)
List::BinarySearch::XS dependency cannot be built on a target system.  It is
provided by the L<List::BinarySearch> distribution.

Examples:


    use List::BinarySearch qw( binsearch  binsearch_pos  binsearch_range );

    # Find the lowest index of a matching element.
    $index = binsearch {$a <=> $b} 300, @{[ 100, 200, 300, 400 ]};
    $index = binsearch {$a cmp $b} 'Mozart', @{[ qw/ Bach Brahms Mozart / ]};
    $index = binsearch {$a <=> $b} 42, @{[ 10, 20, 30 ]}      # not found: undef

    # Find the lowest index of a matching element, or best insert point.
    $index = binsearch_pos {$a cmp $b} 'Chopin', @{[ qw/ Bach Brahms Mozart/ ]};  # Insert at [2].
    $index = binsearch_pos {$a <=> $b} 60, @{[ 10, 20, 30, 40, 50, 70 ]}; # Insert at [5].
    $index = binsearch_pos {$a <=> $b} 20, @{[ 10, 20, 30 ]}; # Matched at [1]


=head1 DESCRIPTION

This module is intended to be used by L<List::BinarySearch>, and shouldn't need
to be used directly in user-code.

This module provides pure-Perl implementations of the C<binsearch> and
C<binsearch_pos> functions for use by L<List::BinarySearch>.  Please refer to
the documentation for L<List::BinarySearch> for a full description of those
functions.  What follows is a very brief overview.

These pure-Perl functions will be overridden by XS code when used via
L<List::BinarySearch> if L<List::BinarySearch::XS> is installed (the default,
and recommended). The pure-Perl functions exist as a gracefull downgrade in case
users aren't able to use XS modules.


=head1 EXPORT

List::BinarySearch::PP exports by default C<binsearch> and C<binsearch_pos>.

=head1 SUBROUTINES/METHODS

=head2 binsearch CODE NEEDLE ARRAY_HAYSTACK

    $first_found_ix = binsearch { $a cmp $b } $needle, @haystack;

Uses the supplied code block as a comparator to search for C<$needle> within
C<@haystack>.  If C<$needle> is found, return value will be the lowest index of
a matching element, or C<undef> if the needle isn't found.

=head2 binsearch_pos CODE NEEDLE ARRAY_HAYSTACK

    $first_found_ix = binsearch_pos { $a cmp $b } $needle, @haystack;

Uses the supplied code block as a comparator to search for C<$needle> within
C<@haystack>. If C<$needle> is found, return value will be the lowest index of
a matching element, or the index of the best insertion point for the needle if
it isn't found.


=head1 CONFIGURATION AND ENVIRONMENT

Perl 5.8 or newer required.  This module is part of the L<List::BinarySearch>
distribution, and is intended for use by the C<List::BinarySearch> module.
Though the user interface is unlikely to change, it shouldn't be directly used
by code outside of this distribution.


=head1 DEPENDENCIES

Perl 5.8.


=head1 INCOMPATIBILITIES

Perl versions prior to 5.8 aren't supported by this distribution.  See the
POD from L<List::BinarySearch> for a more detailed explanation.


=head1 AUTHOR

David Oswald, C<< <davido at cpan.org> >>

If the documentation fails to answer your question, or if you have a comment
or suggestion, send me an email.


=head1 DIAGNOSTICS


=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to
L<https://github.com/daoswald/List-BinarySearch/issues>.  I will
be notified, and then you'll automatically be notified of progress on your bug
as I make changes.



=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc List::BinarySearch

This module is maintained in a public repo at Github.  You may look for
information at:

=over 4

=item * Github: Development is hosted on Github at:

L<http://www.github.com/daoswald/List-BinarySearch>

=item * GitHub Issue tracker (report bugs here)

L<https://github.com/daoswald/List-BinarySearch/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/List-BinarySearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/List-BinarySearch>

=item * Search CPAN

L<http://search.cpan.org/dist/List-BinarySearch/>

=back


=head1 ACKNOWLEDGEMENTS

L<Mastering Algorithms with Perl|http://shop.oreilly.com/product/9781565923980.do>,
from L<O'Reilly|http://www.oreilly.com>: much of the code behind the positional
search.

=head1 LICENSE AND COPYRIGHT

Copyright 2013 David Oswald.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


Makefile.PL
## no critic (RCS,VERSION)

use strict;
use warnings;

use ExtUtils::MakeMaker;

## no critic (may require interpolation eval)

use constant DIST_VERSION => '0.24';

my $prereq_pm = { 'Scalar::Util' => '0' };

if( ! eval 'use List::BinarySearch::XS; 1;' ) {
  my $answer = 'n';
  unless( $ENV{LBS_NO_XS} ) {
    $answer =
      prompt <<'EOP' . 'Would you like to install List::BinarySearch::XS?', 'y';

List::BinarySearch::XS is not installed.

This module will provide a performance improvement.  Though it is not required
for proper operation, it is strongly recommended that this module be installed
unless the target environment is unable to compile XS modules.

Answering 'yes' to the following question will cause most CPAN installers to
pull in the List::BinarySearch::XS module automatically.  If you are doing a
manual installation you will probably need to install List::BinarySearch::XS
separately.

EOP
  }
  if( $answer =~ m/^y/i ) {
    $prereq_pm->{'List::BinarySearch::XS'} = '0'; # Add List::BinarySearch::XS
                                                  # to our list of dependencies.
    print "List::BinarySearch::XS has been added to the dependency list.\n";
  }
  else {
    print "List::BinarySearch::XS will not be added to the dependency list.\n";
  }
}
else {
  print "List::BinarySearch::XS has been detected. It will be used automatically.\n";
}


WriteMakefile(
    NAME               => 'List::BinarySearch',
    AUTHOR             => q{David Oswald <davido [@at] cpan [d.o.t] org>},
    VERSION_FROM       => 'lib/List/BinarySearch.pm',
    ABSTRACT_FROM      => 'lib/List/BinarySearch.pm',
    LICENSE            => 'perl',
    MIN_PERL_VERSION   => '5.008000',
    PL_FILES           => {},
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '6.62',    # Core, but updated version.
    },
    BUILD_REQUIRES => {
        'Test::More' => '0.98',             # Core, but updated version.
        'English'    => '0',                # Core.
    },
    PREREQ_PM => $prereq_pm,
    META_MERGE => {
        'meta-spec'   => { version => 2 },
        resources     => {
            bugtracker => {
                web => 'https://github.com/daoswald/List-BinarySearch/issues',
            },
            repository => {
                type  => 'git',
                url   => 'https://github.com/daoswald/List-BinarySearch.git',
                web   => 'https://github.com/daoswald/List-BinarySearch',
            },
        },
        provides      => {
            'List::BinarySearch' => {
                file    => 'lib/List/BinarySearch.pm',
                version => DIST_VERSION,
            },
            'List::BinarySearch::PP' => {
                file    => 'lib/List/BinarySearch/PP.pm',
                version => DIST_VERSION,
            },
        },
        prereqs => {
            runtime => {
                recommends    => {
                    'List::BinarySearch::XS' => '0.06',
                },
            },
        },
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'List-BinarySearch-*' },
);
